Use properties file for environmental variables

kbrock opened the pull request:
### dependency

- https://github.com/ManageIQ/manageiq-rpm_build/pull/145
- https://github.com/ManageIQ/bundler-inject/pull/17

### purpose

move properties into a properties file to work better with systemd.

### priority

We do not need to make this change and can continue to use our current process. But the format in the 20 or so systemd.service files are a little unwieldy. This new format is available to shell scripts and systemd.

We are changing all these services to work with non-root services, so it makes sense to change all of these at this time.

### what is happening

This change converts our scripts to properties files

After this change, `source /etc/defaults/evm` still sets and exports the same variables to the same values as before

After this change, `HOME` is not longer set to the root user. This caused some issues with the bundler-inject gem but a fix has been proposed to fix this issue.

Fryguy commented on LINK/etc/default/evm:
```diff
@@ -2,23 +2,15 @@
-source /etc/default/evm_production
+for prop in /etc/default/evm.properties /etc/default/evm_production ; do
+  if [[ -f ${prop} ]] ; then 
+    # lines starting with a semi or a # are ignored
```
why semicolon?  is that some standard.  also for readability

```suggestion
    # lines starting with a ';' or a '#' are ignored
```

agrare replied:
This comes from the format of an EnvironmentFile: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#EnvironmentFile=

> Empty lines, lines without an "=" separator, or lines starting with ; or # will be ignored, which may be used for commenting.

agrare replied:
I'm surprised there isn't a tool that does this already...

kbrock commented:
Current issues:

@agrare it is:

- `/etc/default/evm` (path modifications) [ref](https://github.com/ManageIQ/manageiq-appliance/blob/env-props/LINK/etc/default/evm)
  - `/etc/default/evm_ruby` (easy) [ref](https://github.com/ManageIQ/manageiq-appliance/blob/master/LINK/etc/default/evm_ruby)
  - `/etc/default/evm_postgres` (easy)[ref](https://github.com/ManageIQ/manageiq-appliance/blob/master/LINK/etc/default/evm_postgres)
  - `/etc/default/evm_production` (easy) [ref](https://github.com/ManageIQ/manageiq-rpm_build/blob/master/rpm_spec/manageiq.spec.in#L58)
    - `/opt/manageiq/manageiq-gemset/enable`(path modifications)  [built here](https://github.com/ManageIQ/manageiq-rpm_build/blob/master/rpm_spec/manageiq.spec.in#L50-L62)

As you can see, extracting out the properties is trivial. I felt that having a single property file was easiest. Not sure what the separate files bough us.

Also not sure if we should set the `GEM_PATH` and those or if it belongs in a ruby configuration somewhere

agrare commented:
Ah okay `/opt/manageiq/manageiq-gemset/enable` was one that I missed

kbrock commented:
We are allowed to have multiple env properties files. I feel we could cut them up into one per rpm, or just keep as a single file.
Some of these files live here and other live with the rpm build process. They all are appliance and operating system specific. They are also not very dynamic.

Anything that is application specific (used by appliance console) feels like we don't even want a bunch of it and is out of scope for us anyway. Would be nice to just remove the functionality.

I propose we move all properties into a single static file here or to rpm-build, hardcoding the values.
@agrare runs debian so would have the best insight in terms of whether keeping this dynamically created in an rpm buys us anything. @bdunne probably have the most knowledge around this

|file|concern|action|comment|
|---|---------|------|---------|
|[`/etc/default/evm`](https://github.com/ManageIQ/manageiq-appliance/blob/master/LINK/etc/default/evm)|app|trim|keep as a script that sources all properties files - backwards compatibility|
|[`evm.properties`](https://github.com/ManageIQ/manageiq-appliance/blob/master/LINK/etc/default/evm)|app |new|define full hardcoded path|
|[`evm_ruby`](https://github.com/ManageIQ/manageiq-appliance/blob/master/LINK/etc/default/evm_ruby)|app |move into `evm.properties`|everything is just app memory specific values|
|[`evm_postgres`](https://github.com/ManageIQ/manageiq-appliance/blob/master/LINK/etc/default/evm_postgres)|appliance|move to `evm.properties`|most of these could be deleted along with the code that uses them.|
|[`evm_production`](https://github.com/ManageIQ/manageiq-rpm_build/blob/master/rpm_spec/manageiq.spec.in#L58)|appliance console|`evm.properties`|this links to gems-pending and appliance console.|
|[`gemset/enable`](https://github.com/ManageIQ/manageiq-rpm_build/blob/master/rpm_spec/manageiq.spec.in#L50-L62)|appliance (rpm)|keep as `evm.properties`|knows app root and the fact that it is an appliance.|

agrare commented:
> @agrare runs debian so would have the best insight in terms of whether keeping this dynamically created in an rpm buys us anything.

I wouldn't worry about this, these files are appliance specific and our appliance is RHEL/CentOS.

If for some reason we did end up doing a `manageiq-deb_build` we might need to worry about different property files then.

kbrock committed:
don't configure HOME or /usr/local/bin

this needs to work across users, so we may not have access to root's HOME

/usr/local/bin and /usr/local/sbin are already in the path.
since it is tricky to extend values in properties, going the fully hardcoded route

kbrock committed:
move evm_postgres to miq.properties

postgres parameters are appliance specific
they help with PgAdmin and the appliance console

kbrock committed:
move ruby_properties to miq.properties

On our appliance, we have a few settings that make ruby run better.
When our application changes enough, we will probably need to update this.
For that reason, the instructions on how this was derived are included

kbrock committed:
move BUNDLE_GEMFILE from evm_production and /enable to miq.properties

The gemfile is application specific and not ruby specific

kbrock commented:
Upfront: the intent was not performance. But since we had that concern I wanted to measure before and after - especially since we are shelling out to sed and reading an external properties file

Most(all) of the savings was from avoiding the `ruby env` call.

run via `time sh /etc/default/evm`

|type|before|after|
|---|---|---|
|real|0m2.338s|0m0.010s|
|user|0m1.842s|0m0.007s|
|sys|0m0.300s|0m0.002s|

Fryguy commented on LINK/etc/default/evm.properties:
```diff
@@ -0,0 +1,47 @@
+APPLIANCE_TEMPLATE_DIRECTORY=/opt/manageiq/mangeiq-appliance/TEMPLATE
+BUNDLE_GEMFILE=/var/www/miq/vmdb/Gemfile
+GEM_HOME=/opt/manageiq/manageiq-gemset
+GEM_PATH=/opt/manageiq/manageiq-gemset:/usr/share/gems:/usr/local/share/gems
```
The /opt/manageiq is productizable, so it can't be hardcoded in this way.

Fryguy replied:
We might be able to make the rpm emit this file.  Or perhaps we keep it as a template in the template dir and generate from these based on actual system values?

cc @bdunne

kbrock replied:
Cool. I'll put this section back into the rpm.
Will need to make sure this changes in all appliance rpms going forward.

Fryguy reviewed (changes requested):
We can't hardcode /opt/manageiq, so we have to find another way.

kbrock commented:
> We can't hardcode /opt/manageiq, so we have to find another way.
> -- @Fryguy 

Moved that back into the rpm process and setting that there.
If we have other values to use there, we can handle that in other rpms.

using 2 properties files for now.

that is the change I made here and in the rpm build process

bdunne commented on LINK/etc/profile.d/evm.sh:
```diff
@@ -1,5 +1,6 @@
 #!/bin/bash
 
+# note: this will override $PATH
```
Will it?

kbrock replied:
the properties file for `enable` (now called `production.properties`) sets a full path

Fryguy commented:
Sorry to be a pain, but I've been trying to avoid the `miq` short form except where it's required - can we call it `manageiq.properties` ?

kbrock committed:
move properties from evm into miq.properties

properties are going into .properties files to be systemd and shell friendly

removing the sourcing of shell scripts.
each of those files will be added to miq.properties in the next few commits
